# Discovering Disease Outbreaks from News Headlines

Identifying and mapping epidemics is crucial to prevent or respond to deadly disease outbreaks. Your first assignment for the WHO is as follows:

- Extract the locations (city and/or country name) from each news headline.
- Find the geographic coordinates of each headline using the city/country.
- Cluster (group) the headlines based on the geographic location.
- Visualize the clusters on a map and analyze them for patterns indicating an epidemic.
- Investigate the largest clusters for signs of disease outbreaks.
- Review headlines in the largest clusters within the United States and around the world. If any disease outbreak is   particularly dominant, visualize all worldwide mentions of that disease.
- Provide a summary of your findings to your superiors at the WHO so they can direct resources.

# 1. Parsing the News Headlines

**Objective**

Find any city and/or country names mentioned in each of the news headlines.

**Workflow**

1. Load in the headline data and examine it for any data quality issues.
1. Use any library/data structure to read in the headlines
1. Read through some of the headlines and identify potential problems
1. Using regular expressions and the cities and countries within the geonamescache library, match any cities/countries within each headline.
1. Make sure to normalize headlines and city/country names by removing accent marks. This can be done with the unidecode library.
1. Watch out for multiple cities in a headline and matches on short words! We want the match to be on the entire city—for example San Marino—and not a partial match—San.
1. Put the extracted data into a pandas DataFrame with three columns: headline, city, country.
1. Make sure there were no issues with the extraction by sampling some of the headlines and examining the city and country names.
1. One method for finding problems is to look for the most common names and see if there are any issues.
1. Once you are confident you’ve found all the cities/countries in each headline, save the DataFrame for the next part.

**Importance to project**

* We can’t do much with just the headlines; although they contain the city/country names, they do not contain the geographic information—latitude and longitude—we need to find clusters of disease outbreaks. The first step in getting the geographic information is to isolate the names.

* Later, we will use the names to find the location of each headline, which requires bringing in external data (through geonamescache).

* This workflow is common in data science. First, we separate the useful information from the noise—data mining—and then we augment it with external data—data engineering.

## Import all relevent libraries

In [1]:
# Regular expression
import re
from typing import Optional

# Data analysis and wrangling
import numpy as np
import pandas as pd

# Set display options for pandas
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
pd.set_option("max_colwidth", None)

# Normalized unicode data (to remove accents)
import unidecode

## Visualization
# matplotlib
import matplotlib.pyplot as plt

# Ignore warning
import warnings

warnings.filterwarnings("ignore")

## GeonamesCache data


In [2]:
from geonamescache import GeonamesCache

"""
get_continents()
get_countries()
get_us_states()
get_cities()
get_countries_by_names()
get_us_states_by_names()
get_cities_by_name(name)
get_us_counties()
"""
gc = GeonamesCache()

### US states and counties 

#### US states

In [3]:
states = pd.DataFrame(gc.get_us_states_by_names()).T.reset_index(drop=True)
states = states.sort_values(by="name", ascending=False).reset_index(drop=True)
states = states.rename(
    columns={"name": "statename", "code": "statecode", "geonameid": "stateid"}
)
# states.info()
print(states.shape)
states.head()

(51, 4)


statecode      statename fips  stateid
0        WY        Wyoming   56  5843591
1        WI      Wisconsin   55  5279468
2        WV  West Virginia   54  4826850
3        WA     Washington   53  5815135
4        VA       Virginia   51  6254928

> 📝 **Note**: `states` dataFrame consits of 50 states and 1 district of Columbia

#### US Counties

In [4]:
# Retrive countynames data and create a dataframe.
counties = pd.DataFrame(gc.get_us_counties()).T.reset_index(drop=True).T

counties.columns = ["countyid", "countyname", "statecode"]
counties = counties.sort_values(
    by=["statecode", "countyname"], ascending=False
).reset_index(drop=True)
print(counties.shape)
counties.head(10)

(3235, 3)


countyid         countyname statecode
0    56045      Weston County        WY
1    56043    Washakie County        WY
2    56041       Uinta County        WY
3    56039       Teton County        WY
4    56037  Sweetwater County        WY
5    56035    Sublette County        WY
6    56033    Sheridan County        WY
7    56031      Platte County        WY
8    56029        Park County        WY
9    56027    Niobrara County        WY

In [5]:
3225 - 3143

82

> 📝 **Note**: `counties` dataFrame consists of 3,225 counties in the US and its territories. The dataframe has 6 more statecodes and 92 more counies of the US territories. 

In [6]:
counties.statecode.value_counts()

statecode
TX    254
GA    159
VA    134
KY    120
MO    115
KS    105
IL    102
NC    100
IA     99
TN     95
NE     93
IN     92
OH     88
MN     87
MI     83
MS     82
PR     78
OK     77
AR     75
WI     72
AL     67
FL     67
PA     67
SD     66
LA     64
CO     64
NY     62
CA     58
MT     56
WV     55
ND     53
SC     46
ID     44
WA     39
OR     36
NM     33
AK     29
UT     29
MD     24
WY     23
NJ     21
NV     17
ME     16
AZ     15
VT     14
MA     14
NH     10
CT      8
RI      5
HI      5
AS      5
MP      4
VI      3
DE      3
GU      1
DC      1
UM      1
Name: count, dtype: int64

In [7]:
states.statecode.nunique()

51

In [8]:
counties.statecode.nunique()

57

#### Merge US states and counties

In [9]:
# merge state and county dataframes

us_counties = (
    pd.merge(
        counties, states, on="statecode", how="left", suffixes=("_county", "_state")
    )
    .sort_values(by=["statename", "countyname"], ascending=False)
    .reset_index(drop=True)
)
us_counties = us_counties.rename(columns={"countyname": "county", "statename": "state"})

# Remove accents from county names
us_counties["county"] = us_counties["county"].apply(lambda x: unidecode.unidecode(x))

print(us_counties.shape)
us_counties.head()

(3235, 6)


countyid             county statecode    state fips  stateid
0    56045      Weston County        WY  Wyoming   56  5843591
1    56043    Washakie County        WY  Wyoming   56  5843591
2    56041       Uinta County        WY  Wyoming   56  5843591
3    56039       Teton County        WY  Wyoming   56  5843591
4    56037  Sweetwater County        WY  Wyoming   56  5843591

In [10]:
# How many counties and states that have no state code in `states` dataframe?
us_counties[us_counties.state.isna()]

countyid                         county statecode state fips stateid
3143    72153                Yauco Municipio        PR   NaN  NaN     NaN
3144    72151              Yabucoa Municipio        PR   NaN  NaN     NaN
3145    60050               Western District        AS   NaN  NaN     NaN
3146    72149             Villalba Municipio        PR   NaN  NaN     NaN
3147    72147              Vieques Municipio        PR   NaN  NaN     NaN
3148    72145            Vega Baja Municipio        PR   NaN  NaN     NaN
3149    72143            Vega Alta Municipio        PR   NaN  NaN     NaN
3150    72141               Utuado Municipio        PR   NaN  NaN     NaN
3151    72139        Trujillo Alto Municipio        PR   NaN  NaN     NaN
3152    72137             Toa Baja Municipio        PR   NaN  NaN     NaN
3153    72135             Toa Alta Municipio        PR   NaN  NaN     NaN
3154    69120            Tinian Municipality        MP   NaN  NaN     NaN
3155    60040                  Swains Island        AS   NaN  NaN     NaN
3156    78030              St. Thomas Island        VI   NaN  NaN     NaN
3157    78020                St. John Island        VI   NaN  NaN     NaN
3158    78010               St. Croix Island        VI   NaN  NaN     NaN
3159    72133         Santa Isabel Municipio        PR   NaN  NaN     NaN
3160    72131        San Sebastian Municipio        PR   NaN  NaN     NaN
3161    72129          San Lorenzo Municipio        PR   NaN  NaN     NaN
3162    72127             San Juan Municipio        PR   NaN  NaN     NaN
3163    72125           San German Municipio        PR   NaN  NaN     NaN
3164    72123              Salinas Municipio        PR   NaN  NaN     NaN
3165    69110            Saipan Municipality        MP   NaN  NaN     NaN
3166    72121        Sabana Grande Municipio        PR   NaN  NaN     NaN
3167    69100              Rota Municipality        MP   NaN  NaN     NaN
3168    60030                    Rose Island        AS   NaN  NaN     NaN
3169    72119           Rio Grande Municipio        PR   NaN  NaN     NaN
3170    72117               Rincon Municipio        PR   NaN  NaN     NaN
3171    72115         Quebradillas Municipio        PR   NaN  NaN     NaN
3172    72113                Ponce Municipio        PR   NaN  NaN     NaN
3173    72111             Penuelas Municipio        PR   NaN  NaN     NaN
3174    72109             Patillas Municipio        PR   NaN  NaN     NaN
3175    72107             Orocovis Municipio        PR   NaN  NaN     NaN
3176    69085  Northern Islands Municipality        MP   NaN  NaN     NaN
3177    72105            Naranjito Municipio        PR   NaN  NaN     NaN
3178    72103              Naguabo Municipio        PR   NaN  NaN     NaN
3179    72101              Morovis Municipio        PR   NaN  NaN     NaN
3180    72099                 Moca Municipio        PR   NaN  NaN     NaN
3181    74300                 Midway Islands        UM   NaN  NaN     NaN
3182    72097             Mayaguez Municipio        PR   NaN  NaN     NaN
3183    72095              Maunabo Municipio        PR   NaN  NaN     NaN
3184    72093              Maricao Municipio        PR   NaN  NaN     NaN
3185    60020                Manu'a District        AS   NaN  NaN     NaN
3186    72091               Manati Municipio        PR   NaN  NaN     NaN
3187    72089             Luquillo Municipio        PR   NaN  NaN     NaN
3188    72087                Loiza Municipio        PR   NaN  NaN     NaN
3189    72085          Las Piedras Municipio        PR   NaN  NaN     NaN
3190    72083           Las Marias Municipio        PR   NaN  NaN     NaN
3191    72081                Lares Municipio        PR   NaN  NaN     NaN
3192    72079                Lajas Municipio        PR   NaN  NaN     NaN
3193    72077               Juncos Municipio        PR   NaN  NaN     NaN
3194    72075           Juana Diaz Municipio        PR   NaN  NaN     NaN
3195    72073               Jayuya Municipio        PR   NaN  NaN     NaN
3196    7

In [11]:
def remove_county_suffix(county: str) -> Optional[str]:
    """Remove common suffixes from county names."""
    s = [
        "County",
        "Municipality",
        "Municipio",
        "Census Area",
        "City and Borough",
        "Borough",
        "Parish",
    ]
    #  Create a regex pattern to match any of the suffixes

    regexs = "|".join(s)

    # Use regex to remove the suffixes
    county_name_only = re.sub(rf"\s*({regexs})\s*$", "", county, flags=re.IGNORECASE)
    return county_name_only.strip() if county_name_only else None


# Insert a new column for cleaned county names
us_counties.insert(loc=2, column="countyname", value=None)

# Apply the function to the 'county' column
us_counties["countyname"] = us_counties["county"].apply(remove_county_suffix)


print(counties.shape)


us_counties.head(10)

(3235, 3)


countyid             county  countyname statecode    state fips  stateid
0    56045      Weston County      Weston        WY  Wyoming   56  5843591
1    56043    Washakie County    Washakie        WY  Wyoming   56  5843591
2    56041       Uinta County       Uinta        WY  Wyoming   56  5843591
3    56039       Teton County       Teton        WY  Wyoming   56  5843591
4    56037  Sweetwater County  Sweetwater        WY  Wyoming   56  5843591
5    56035    Sublette County    Sublette        WY  Wyoming   56  5843591
6    56033    Sheridan County    Sheridan        WY  Wyoming   56  5843591
7    56031      Platte County      Platte        WY  Wyoming   56  5843591
8    56029        Park County        Park        WY  Wyoming   56  5843591
9    56027    Niobrara County    Niobrara        WY  Wyoming   56  5843591

### Countries and cities


#### Countries

In [12]:
countries = pd.DataFrame(gc.get_countries_by_names()).T
# sort countries by name and reset index
countries = countries.sort_values(by="name", ascending=True).reset_index(drop=True)
# Rename columns for clarity
countries.rename(
    columns={"iso": "countrycode", "name": "countryname", "geonameid": "countryid"},
    inplace=True,
)

# countries.info()
print(countries.shape)
countries.head(10)

(252, 17)


countryid          countryname countrycode iso3 isonumeric fips  \
0   1149361          Afghanistan          AF  AFG          4   AF   
1    661882        Aland Islands          AX  ALA        248        
2    783754              Albania          AL  ALB          8   AL   
3   2589581              Algeria          DZ  DZA         12   AG   
4   5880801       American Samoa          AS  ASM         16   AQ   
5   3041565              Andorra          AD  AND         20   AN   
6   3351879               Angola          AO  AGO         24   AO   
7   3573511             Anguilla          AI  AIA        660   AV   
8   6697173           Antarctica          AQ  ATA         10   AY   
9   3576396  Antigua and Barbuda          AG  ATG         28   AC   

  continentcode           capital   areakm2 population  tld currencycode  \
0            AS             Kabul    647500   37172386  .af          AFN   
1            EU         Mariehamn      1580      26711  .ax          EUR   
2            EU            Tirana     28748    2866376  .al          ALL   
3            AF           Algiers   2381740   42228429  .dz          DZD   
4            OC         Pago Pago       199      55465  .as          USD   
5            EU  Andorra la Vella       468      77006  .ad          EUR   
6            AF            Luanda   1246700   30809762  .ao          AOA   
7            NA        The Valley       102      13254  .ai          XCD   
8            AN                    14000000          0  .aq                
9            NA        St. John's       443      96286  .ag          XCD   

  currencyname    phone   postalcoderegex          languages  \
0      Afghani       93                    fa-AF,ps,uz-AF,tk   
1         Euro  +358-18  ^(?:FI)*(\d{5})$              sv-AX   
2          Lek      355         ^(\d{4})$              sq,el   
3        Dinar      213         ^(\d{5})$              ar-DZ   
4       Dollar   +1-684             96799        en-AS,sm,to   
5         Euro      376  ^(?:AD)*(\d{3})$                 ca   
6       Kwanza      244                                pt-AO   
7       Dollar   +1-264                                en-AI   
8                                                              
9       Dollar   +1-268                                en-AG   

             neighbours  
0     TM,CN,IR,TJ,PK,UZ  
1                        
2        MK,GR,ME,RS,XK  
3  NE,EH,LY,MR,TN,MA,ML  
4                        
5                 ES,FR  
6           CD,NA,ZM,CG  
7                        
8                        
9

In [13]:
# Keep only nessessary columns
countries = countries[["countryname", "capital", "countrycode", "countryid"]]

countries.head(10)

countryname           capital countrycode countryid
0          Afghanistan             Kabul          AF   1149361
1        Aland Islands         Mariehamn          AX    661882
2              Albania            Tirana          AL    783754
3              Algeria           Algiers          DZ   2589581
4       American Samoa         Pago Pago          AS   5880801
5              Andorra  Andorra la Vella          AD   3041565
6               Angola            Luanda          AO   3351879
7             Anguilla        The Valley          AI   3573511
8           Antarctica                            AQ   6697173
9  Antigua and Barbuda        St. John's          AG   3576396

#### Cities


In [14]:
# create a dataframe for cities
cities = pd.DataFrame(gc.get_cities()).T.reset_index(drop=True)
print(cities.shape)
cities.head()

(26463, 9)


geonameid                 name  latitude longitude countrycode population  \
0   3040051         les Escaldes  42.50729   1.53414          AD      15853   
1   3041563     Andorra la Vella  42.50779   1.52109          AD      20430   
2    290594   Umm Al Quwain City  25.56473  55.55517          AE      62747   
3    291074  Ras Al Khaimah City  25.78953   55.9432          AE     351943   
4    291580           Zayed City  23.65416  53.70522          AE      63482   

         timezone admin1code  \
0  Europe/Andorra         08   
1  Europe/Andorra         07   
2      Asia/Dubai         07   
3      Asia/Dubai         05   
4      Asia/Dubai         01   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  alternatenames  
0                                                                                                                                                                                                                                                                                                        [Ehskal'des-Ehndzhordani, Escaldes, Escaldes-Engordany, Les Escaldes, esukarudesu=engorudani jiao qu, lai sai si ka er de-en ge er da, Эскальдес-Энджордани, エスカルデス＝エンゴルダニ教区, 萊塞斯卡爾德-恩戈爾達, 萊塞斯卡爾德－恩戈爾達]  
1  [ALV, Ando-la-Vyey, Andora, Andora la Vela, Andora la Velja, Andora lja Vehl'ja, Andoro Malnova, Andorra, Andorra Tuan, Andorra a Vella, Andorra la Biella, Andorra la Vella, Andorra la Vielha, Andorra-a-Velha, Andorra-la-Vel'ja, Andorra-la-Vielye, Andorre-la-Vieille, Andò-la-Vyèy, Andòrra la Vièlha, an dao er cheng, andolalabeya, andwra la fyla, Ανδόρρα, Андора ла Веля, Андора ла Веља, Андора ля Вэлья, Андорра-ла-Велья, אנדורה לה וולה, أندورا لا فيلا, አንዶራ ላ ቬላ, アンドラ・ラ・ヴェリャ, 安道爾城, 안도라라베야]  
2                                                                                                                                                                                                                                               [Oumm al Qaiwain, Oumm al Qaïwaïn, Um al Kawain, Um al Quweim, Umm Al Quwain City, Umm al Qaiwain, Umm al Qawain, Umm al Qaywayn, Umm al-Quwain, Umm-ehl'-Kajvajn, Yumul al Quwain, am alqywyn, mdynt am alqywyn, Умм-эль-Кайвайн, أم القيوين, مدينة ام القيوين]  
3                                                                                                                                                                                                            [Julfa, Khaimah, RAK City, RKT, Ra's al Khaymah, Ra's al-Chaima, Ras Al Khaimah City, Ras al Khaimah, Ras al-Khaimah, Ras el Khaimah, Ras el Khaïmah, Ras el-Kheima, Ras-ehl'-Khajma, Ra’s al Khaymah, Ra’s al-Chaima, mdynt ras alkhymt, ras alkhymt, Рас-эль-Хайма, رأس الخيمة, مدينة رأس الخيمة]  
4                                                                                                                                                                                                                                                                                                                                                                                       [Bid' Zayed, Bid’ Zayed, Madinat Za'id, Madinat Zayid, Madīnat Zāyid, Madīnat Zā’id, Zayed City, mdynt zayd, مدينة زايد]

In [15]:
# sort cities by country code and name in descending order
cities = cities.sort_values(by=["countrycode", "name"]).reset_index(drop=True)
# drop unnecessary columns
cities = cities.drop(columns=["population", "alternatenames"])
# rename columns for clarity
cities = cities.rename({"geonameid": "cityid", "name": "cityname_raw"}, axis=1)
# transform all accented strings to English alphabets'
cities.insert(loc=2, column="cityname", value=np.nan)
cities.cityname = cities["cityname_raw"].apply(
    func=lambda x: unidecode.unidecode(string=x)
)

cities.head(10)

cityid      cityname_raw          cityname  latitude longitude  \
0   3041563  Andorra la Vella  Andorra la Vella  42.50779   1.52109   
1   3040051      les Escaldes      les Escaldes  42.50729   1.53414   
2    292968         Abu Dhabi         Abu Dhabi  24.45118  54.39696   
3    292953         Adh Dhayd         Adh Dhayd  25.28812  55.88157   
4    292932        Ajman City        Ajman City  25.40177  55.47878   
5    292913       Al Ain City       Al Ain City  24.19167  55.76056   
6    292878  Al Fujairah City  Al Fujairah City  25.11641  56.34141   
7  12047416  Al Shamkhah City  Al Shamkhah City  24.39268  54.70779   
8    292688         Ar Ruways         Ar Ruways  24.11028  52.73056   
9  12042052     Bani Yas City     Bani Yas City  24.30978  54.62944   

  countrycode        timezone admin1code  
0          AD  Europe/Andorra         07  
1          AD  Europe/Andorra         08  
2          AE      Asia/Dubai         01  
3          AE      Asia/Dubai         06  
4          AE      Asia/Dubai         02  
5          AE      Asia/Dubai         01  
6          AE      Asia/Dubai         04  
7          AE      Asia/Dubai         01  
8          AE      Asia/Dubai         01  
9          AE      Asia/Dubai         01

#### Merge `countries` dataframes to `cities` dataframe 

In [16]:
# How many different country codes are there in the cities dataframe?
len(cities.countrycode.unique())

244

> ⚠️ **Caution**: The `countries` and `cities` dataframes do not have the same number of countries. The `countries` dataframe has 252 countries, but `cities` dataframe has only 244 countries. Because of some countries not having cities in the `cities` df, we will use an outer join to merge the two dataframes. This will ensure that all countries are included, even if they do not have any cities listed in the `cities` df.


In [17]:
city = (
    cities[["cityname", "latitude", "longitude", "countrycode", "cityid"]]
    .merge(countries, how="left", on="countrycode", suffixes=("_city", "_country"))
    .sort_values(by=["countrycode", "cityname"], ascending=True)
    .reset_index(drop=True)
)
# rearrange columns for better readability
city = city[
    [
        "cityname",
        "countryname",
        "capital",
        "latitude",
        "longitude",
        "countrycode",
        "cityid",
        "countryid",
    ]
]

print(city.shape)
city.head(10)

(26463, 8)


cityname           countryname           capital  latitude  \
0  Andorra la Vella               Andorra  Andorra la Vella  42.50779   
1      les Escaldes               Andorra  Andorra la Vella  42.50729   
2         Abu Dhabi  United Arab Emirates         Abu Dhabi  24.45118   
3         Adh Dhayd  United Arab Emirates         Abu Dhabi  25.28812   
4        Ajman City  United Arab Emirates         Abu Dhabi  25.40177   
5       Al Ain City  United Arab Emirates         Abu Dhabi  24.19167   
6  Al Fujairah City  United Arab Emirates         Abu Dhabi  25.11641   
7  Al Shamkhah City  United Arab Emirates         Abu Dhabi  24.39268   
8         Ar Ruways  United Arab Emirates         Abu Dhabi  24.11028   
9     Bani Yas City  United Arab Emirates         Abu Dhabi  24.30978   

  longitude countrycode    cityid countryid  
0   1.52109          AD   3041563   3041565  
1   1.53414          AD   3040051   3041565  
2  54.39696          AE    292968    290557  
3  55.88157          AE    292953    290557  
4  55.47878          AE    292932    290557  
5  55.76056          AE    292913    290557  
6  56.34141          AE    292878    290557  
7  54.70779          AE  12047416    290557  
8  52.73056          AE    292688    290557  
9  54.62944          AE  12042052    290557

In [18]:
# city.columns.tolist()

## Headline data


In [19]:
file_contents = ""
with open("headlines.txt", "r") as file_handle:
    for line in file_handle.readlines():
        file_contents += line.replace("-", " ")
        unidecode.unidecode(line)

# View first 500 charactors
file_contents[:500]

'Zika Outbreak Hits Miami\nCould Zika Reach New York City?\nFirst Case of Zika in Miami Beach\nMystery Virus Spreads in Recife, Brazil\nDallas man comes down with case of Zika\nTrinidad confirms first Zika case\nZika Concerns are Spreading in Houston\nGeneve Scientists Battle to Find Cure\nThe CDC in Atlanta is Growing Worried\nZika Infested Monkeys in Sao Paulo\nBrownsville teen contracts Zika virus\nMosquito control efforts in St. Louis take new tactics with Zika threat\nSan Juan reports 1st U.S. Zika rela'

In [26]:
# Create dataframe of headlines
df = (
    pd.DataFrame(file_contents.split("\n"), columns=["headline"])
    .sort_values(by="headline", ascending=True)
    .reset_index(drop=True)
)
print(df.shape)
df.head(10)

(650, 1)


headline
0                                              18 new Zika Cases in Bogota
1                                            19 new Zika Cases in Sengkang
2                                 Alameda Residents Recieve Rabies vaccine
3             Albany Residents Recieve Respiratory Syncytial Virus vaccine
4                                    Antipolo under threat from Zika Virus
5                                        Arhus is infested with Bronchitis
6                                         Arvada is infested with Syphilis
7                                                      Authorities a Miami
8  Authorities are Worried about the Spread of Bronchitis in Silver Spring
9          Authorities are Worried about the Spread of Chickenpox in Hemet

In [27]:
df

headline
0                                                                18 new Zika Cases in Bogota
1                                                              19 new Zika Cases in Sengkang
2                                                   Alameda Residents Recieve Rabies vaccine
3                               Albany Residents Recieve Respiratory Syncytial Virus vaccine
4                                                      Antipolo under threat from Zika Virus
5                                                          Arhus is infested with Bronchitis
6                                                           Arvada is infested with Syphilis
7                                                                        Authorities a Miami
8                    Authorities are Worried about the Spread of Bronchitis in Silver Spring
9                            Authorities are Worried about the Spread of Chickenpox in Hemet
10                        Authorities are Worried about the Spread of Chickenpox in Richmond
11                            Authorities are Worried about the Spread of Dengue in Kingston
12                     Authorities are Worried about the Spread of Gonorrhea in Taoyuan City
13                           Authorities are Worried about the Spread of Hepatitis B in Yiwu
14                          Authorities are Worried about the Spread of Hepatitis D in Akron
15                          Authorities are Worried about the Spread of Hepatitis D in Ganja
16                      Authorities are Worried about the Spread of Hepatitis D in North Bay
17                         Authorities are Worried about the Spread of Influenza in Savannah
18                  Authorities are Worried about the Spread of Mad Cow Disease in Edinburgh
19                       Authorities are Worried about the Spread of Mad Cow Disease in Rome
20                               Authorities are Worried about the Spread of Mumps in Corona
21                     Authorities are Worried about the Spread of Norovirus in Buenos Aires
22                            Authorities are Worried about the Spread of Norovirus in Dubai
23                            Authorities are Worried about the Spread of Rotavirus in Union
24                         Authorities are Worried about the Spread of Syphilis in Pyongyang
25                         Authorities are Worried about the Spread of Tuberculosis in Abuja
26                         Authorities are Worried about the Spread of Tuberculosis in Cantu
27                           Authorities are Worried about the Spread of Varicella in Clovis
28                                                             BREAKING – Zika in Missoula
29                                            Bad Water Leads to Sickness in Flint, Michigan
30                                                            Baltimore plans for Zika virus
31                                                           Barcelona Struck by Spanish Flu
32                                                          Barcelona Struck by Spanish Flu 
33                            Barstow Patient in Critical Condition after Contracting Herpes
34                                             Batangas Tourism Takes a Hit as Virus Spreads
35                                                           Beijing team cracks Zika secret
36                                                        Belize City under threat from Zika
37                                                Benton Residents Recieve Pneumonia vaccine
38                                                Biloxi Residents Recieve Chlamydia vaccine
39                                               Birmingham Residents Recieve Rabies vaccine
40                                                     Bradenton Experiences Zika Troubles  
41                            Bridgeport authorities confirmed the spread of West Nile Virus
42                                                      Brisbane is infested with Men

### Removes duplicates

In [28]:
# Check for duplicate headlines
df[df.headline.duplicated(False)]

headline
421        Spanish Flu Outbreak in Lisbon
422        Spanish Flu Outbreak in Lisbon
424  Spanish Flu Spreading through Madrid
425  Spanish Flu Spreading through Madrid

In [30]:
# remove duplicates
df.drop_duplicates(subset="headline", inplace=True, ignore_index=True)
print(df.shape)

(648, 1)


### Add headline IDs

In [ ]:
#  use index number as headlineid
df.reset_index(drop=False, names="headlineid", inplace=True)

In [45]:
df.headlineid = df.headlineid.astype(str)
# Zero pad headlineid to 3 digits
df.headlineid = df.headlineid.str.zfill(3)

In [46]:
df

headlineid  \
0          000   
1          001   
2          002   
3          003   
4          004   
5          005   
6          006   
7          007   
8          008   
9          009   
10         010   
11         011   
12         012   
13         013   
14         014   
15         015   
16         016   
17         017   
18         018   
19         019   
20         020   
21         021   
22         022   
23         023   
24         024   
25         025   
26         026   
27         027   
28         028   
29         029   
30         030   
31         031   
32         032   
33         033   
34         034   
35         035   
36         036   
37         037   
38         038   
39         039   
40         040   
41         041   
42         042   
43         043   
44         044   
45         045   
46         046   
47         047   
48         048   
49         049   
50         050   
51         051   
52         052   
53         053   
54         054   
55         055   
56         056   
57         057   
58         058   
59         059   
60         060   
61         061   
62         062   
63         063   
64         064   
65         065   
66         066   
67         067   
68         068   
69         069   
70         070   
71         071   
72         072   
73         073   
74         074   
75         075   
76         076   
77         077   
78         078   
79         079   
80         080   
81         081   
82         082   
83         083   
84         084   
85         085   
86         086   
87         087   
88         088   
89         089   
90         090   
91         091   
92         092   
93         093   
94         094   
95         095   
96         096   
97         097   
98         098   
99         099   
100        100   
101        101   
102        102   
103        103   
104        104   
105        105   
106        106   
107        107   
108        108   
109        109   
110        110   
111        111   
112        112   
113        113   
114        114   
115        115   
116        116   
117        117   
118        118   
119        119   
120        120   
121        121   
122        122   
123        123   
124        124   
125        125   
126        126   
127        127   
128        128   
129        129   
130        130   
131        131   
132        132   
133        133   
134        134   
135        135   
136        136   
137        137   
138        138   
139        139   
140        140   
141        141   
142        142   
143        143   
144        144   
145        145   
146        146   
147        147   
148        148   
149        149   
150        150   
151        151   
152        152   
153        153   
154        154   
155        155   
156        156   
157        157   
158        158   
159        159   
160        160   
161        161   
162        162   
163        163   
164        164   
165        165   
166        166   
167        167   
168        168   
169        169   
170        170   
171        171   
172        172   
173        173   
174        174   
175        175   
176        176   
177        177   
178        178   
179        179   
180        180   
181        181   
182        182   
183        183   
184        184   
185        185   
186        186   
187        187   
188        188   
189        189   
190        190   
191        191   
192        192   
193        193   
194        194   
195        195   
196        196   
197        197   
198        198   
199        199   
200        200   
201        201   
202        202   
203        203   
204        204   
205        205   
206        206   
207        207   
208        208   
209        209   
210        210   
211        211   
212        212   
213        213   
214        214   
215        215   
216        216   
217        217   
218        218   
219        219   
220        220   
221     

## Extract cities and countries from headlines

### Use regex to retrive all country and city names from headline column

In [47]:
from re import Pattern
from typing import Any, Optional, Iterable


# def get_compiled_regex(name: str) -> Pattern[Any]:
#     """Create a regex pattern for the given name. Compiled one name at a time."""
#     regex = f"\\b{name}\\b"
#     return re.compile(regex)  # re.IGNORECASE)


def get_big_compiled_regex(namelist: Iterable) -> Pattern[Any]:
    """
    Create a regex pattern for the given name list.
    Ensures input is a list, sorts by length (descending), then compiles.
    """
    if not isinstance(namelist, list):
        namelist = list(namelist)
    namelist.sort(key=len, reverse=True)
    all_names = "|".join(re.escape(name) for name in namelist)
    regex = f"\\b(?:{all_names})\\b"
    # return re.compile(regex, re.IGNORECASE)
    return re.compile(regex)

In [48]:
# def get_matched_country(headline: str) -> Optional[str]:
#     """Extract the matched country string from a headline string by compiled one regex pattern of country at a time."""
#     # Create a compiled regex for each country name
#     for country_name in city.countryname:
#         compiled_country = get_compiled_regex(country_name)
#         # match country in headline
#         match_country = compiled_country.search(headline)
#         if match_country is not None:
#             start, end = match_country.start(), match_country.end()
#             return headline[start:end]


def get_matched_country(headline: str) -> Optional[str]:
    """Extract the matched country name string from a headline string by compiled regex pattern of all countries."""
    compiled_countries = get_big_compiled_regex(city.countryname)
    matched_countries = re.findall(compiled_countries, headline)
    return matched_countries[0] if matched_countries else None


def get_matched_city(headline: str) -> Optional[str]:
    """Extract the matched city name string from a headline string by compiled regex pattern of all cities."""
    compiled_cities = get_big_compiled_regex(city.cityname)
    matched_cities = re.findall(compiled_cities, headline)
    return matched_cities[0] if matched_cities else None


def get_matched_usstate(headline: str) -> Optional[str]:
    """Extract the matched US state name string from a headline string by compiled regex pattern of all US states."""
    compiled_states = get_big_compiled_regex(states.statename)
    matched_states = re.findall(compiled_states, headline)
    return matched_states[0] if matched_states else None


def get_matched_uscounty(headline: str) -> Optional[str]:
    """Extract the matched US county name string from a headline string by compiled regex pattern of all US counties."""
    compiled_counties = get_big_compiled_regex(us_counties.countyname)
    matched_counties = re.findall(compiled_counties, headline)
    return matched_counties[0] if matched_counties else None

    # """Extract the matched city string from a regex match object."""
    # compiled_city = get_city_regex(city_name)
    # match_city = compiled_city.search(headline)
    # if match_city is not None:
    #     start, end = match_city.start(), match_city.end()
    #     return headline[start:end]

### Cities extraction


In [49]:
df["city"] = df.headline.apply(func=lambda x: get_matched_city(x))

print(df.shape)
df.head(10)

(648, 3)


headlineid  \
0        000   
1        001   
2        002   
3        003   
4        004   
5        005   
6        006   
7        007   
8        008   
9        009   

                                                                  headline  \
0                                              18 new Zika Cases in Bogota   
1                                            19 new Zika Cases in Sengkang   
2                                 Alameda Residents Recieve Rabies vaccine   
3             Albany Residents Recieve Respiratory Syncytial Virus vaccine   
4                                    Antipolo under threat from Zika Virus   
5                                        Arhus is infested with Bronchitis   
6                                         Arvada is infested with Syphilis   
7                                                      Authorities a Miami   
8  Authorities are Worried about the Spread of Bronchitis in Silver Spring   
9          Authorities are Worried about the Spread of Chickenpox in Hemet   

            city  
0         Bogota  
1       Sengkang  
2        Alameda  
3         Albany  
4       Antipolo  
5          Arhus  
6         Arvada  
7          Miami  
8  Silver Spring  
9          Hemet

> 📝 **Note**:
> The `cities` extraction is done using regex to match the city names in the headlines. The regex pattern and matching sequence (lengthiest name first) is designed to capture full city names, avoiding partial matches. 
>
> - of total 648 headlines, 
>   - 606 headlines (93.5%) have city names extracted.
>   - 42 headlines (6.5%) do not have any city names extracted.    





In [52]:
df

headlineid  \
0          000   
1          001   
2          002   
3          003   
4          004   
5          005   
6          006   
7          007   
8          008   
9          009   
10         010   
11         011   
12         012   
13         013   
14         014   
15         015   
16         016   
17         017   
18         018   
19         019   
20         020   
21         021   
22         022   
23         023   
24         024   
25         025   
26         026   
27         027   
28         028   
29         029   
30         030   
31         031   
32         032   
33         033   
34         034   
35         035   
36         036   
37         037   
38         038   
39         039   
40         040   
41         041   
42         042   
43         043   
44         044   
45         045   
46         046   
47         047   
48         048   
49         049   
50         050   
51         051   
52         052   
53         053   
54         054   
55         055   
56         056   
57         057   
58         058   
59         059   
60         060   
61         061   
62         062   
63         063   
64         064   
65         065   
66         066   
67         067   
68         068   
69         069   
70         070   
71         071   
72         072   
73         073   
74         074   
75         075   
76         076   
77         077   
78         078   
79         079   
80         080   
81         081   
82         082   
83         083   
84         084   
85         085   
86         086   
87         087   
88         088   
89         089   
90         090   
91         091   
92         092   
93         093   
94         094   
95         095   
96         096   
97         097   
98         098   
99         099   
100        100   
101        101   
102        102   
103        103   
104        104   
105        105   
106        106   
107        107   
108        108   
109        109   
110        110   
111        111   
112        112   
113        113   
114        114   
115        115   
116        116   
117        117   
118        118   
119        119   
120        120   
121        121   
122        122   
123        123   
124        124   
125        125   
126        126   
127        127   
128        128   
129        129   
130        130   
131        131   
132        132   
133        133   
134        134   
135        135   
136        136   
137        137   
138        138   
139        139   
140        140   
141        141   
142        142   
143        143   
144        144   
145        145   
146        146   
147        147   
148        148   
149        149   
150        150   
151        151   
152        152   
153        153   
154        154   
155        155   
156        156   
157        157   
158        158   
159        159   
160        160   
161        161   
162        162   
163        163   
164        164   
165        165   
166        166   
167        167   
168        168   
169        169   
170        170   
171        171   
172        172   
173        173   
174        174   
175        175   
176        176   
177        177   
178        178   
179        179   
180        180   
181        181   
182        182   
183        183   
184        184   
185        185   
186        186   
187        187   
188        188   
189        189   
190        190   
191        191   
192        192   
193        193   
194        194   
195        195   
196        196   
197        197   
198        198   
199        199   
200        200   
201        201   
202        202   
203        203   
204        204   
205        205   
206        206   
207        207   
208        208   
209        209   
210        210   
211        211   
212        212   
213        213   
214        214   
215        215   
216        216   
217        217   
218        218   
219        219   
220        220   
221     

In [50]:
print(len(df[df.city.isnull()]))
df[df.city.isnull()]

42


headlineid  \
45         045   
52         052   
72         072   
99         099   
111        111   
113        113   
182        182   
183        183   
207        207   
208        208   
222        222   
223        223   
243        243   
244        244   
265        265   
274        274   
277        277   
280        280   
282        282   
292        292   
328        328   
329        329   
382        382   
393        393   
405        405   
422        422   
430        430   
432        432   
452        452   
460        460   
504        504   
522        522   
526        526   
530        530   
561        561   
573        573   
590        590   
595        595   
597        597   
606        606   
622        622   
639        639   

                                                                              headline  \
45                                                    Bronchitis Outbreak in Manhasset   
52                        Carnival under threat in Rio De Janeiro due to Zika outbreak   
72                                                Chikungunya has not Left Pismo Beach   
99                                         Fort Belvoir tests new cure for Hepatitis C   
111                                              Greenwich Establishes Zika Task Force   
113                   Gympie Patient in Critical Condition after Contracting Chlamydia   
182                                       Longwood volunteers spreading Zika awareness   
183                                                      Louisiana Zika cases up to 26   
207                                           Maka City Experiences Influenza Outbreak   
208                                                         Malaria Exposure in Sussex   
222                                            Martinsville tests new cure for Measles   
223                                                                 Measles Hits Davos   
243                       More Patients in Magnolia are Getting Diagnosed with Malaria   
244                       More Patients in Maynard are Getting Diagnosed with Syphilis   
265                                             More Zika patients reported in Mcallen   
274                             More people in Boucau are infected with HIV every year   
277                           More people in Huron are infected with Dengue every year   
280                     More people in Mclean are infected with Hepatitis A every year   
282  More people in Oak Brook are infected with Respiratory Syncytial Virus every year   
292                                           Mumps Symptoms Spread all over Jefferson   
328                                                         Outbreak of Zika in Destin   
329                                                       Outbreak of Zika in Hutchins   
382                  Rumors about Influenza Spreading in Dobbs Ferry have been Refuted   
393                         Rumors about Syphilis spreading in Penal have been refuted   
405                                 Schools in Coamo Closed Due to Rhinovirus Outbreak   
422                                                     Spanish Flu Sighted in Antigua   
430                                        Spike of Hepatitis E Cases in Morehead City   
432                                           Spike of Meningitis Cases in Druid Hills   
452                                                            Syphilis Hits Jefferson   
460                           The Spread of Chikungunya in Davidson has been Confirmed   
504                                            West Nile Virus Outbreak in Saint Johns   
522                                       Will Tuberculosis vaccine help Cherry Creek?   
526                                         Will West Nile Virus vaccine help Parsons?   
530                                          Yulee takes a hit from Spreading Sickness   
561                                              Zika Troubles come to Padre La

### Add country names column to the DataFrame by merging with `city` dataframe 

In [56]:
merged = df.merge(
    right=city[["countryname", "cityname", "cityid"]],
    how="left",
    left_on="city",
    right_on="cityname",
    suffixes=("_headline", "_city"),
)

merged

headlineid  \
0           000   
1           001   
2           002   
3           003   
4           003   
5           003   
6           003   
7           003   
8           004   
9           005   
10          006   
11          007   
12          008   
13          009   
14          010   
15          010   
16          010   
17          010   
18          010   
19          010   
20          010   
21          010   
22          010   
23          011   
24          011   
25          011   
26          011   
27          012   
28          013   
29          014   
30          015   
31          016   
32          017   
33          018   
34          019   
35          019   
36          019   
37          020   
38          020   
39          021   
40          022   
41          023   
42          024   
43          025   
44          026   
45          027   
46          027   
47          028   
48          029   
49          029   
50          030   
51          031   
52          031   
53          032   
54          032   
55          033   
56          034   
57          035   
58          036   
59          037   
60          038   
61          039   
62          039   
63          039   
64          040   
65          041   
66          041   
67          042   
68          043   
69          044   
70          044   
71          044   
72          044   
73          045   
74          046   
75          046   
76          047   
77          048   
78          048   
79          048   
80          049   
81          050   
82          051   
83          052   
84          053   
85          054   
86          055   
87          055   
88          056   
89          057   
90          057   
91          057   
92          057   
93          057   
94          058   
95          059   
96          059   
97          060   
98          061   
99          061   
100         061   
101         062   
102         063   
103         063   
104         063   
105         063   
106         063   
107         064   
108         064   
109         065   
110         066   
111         066   
112         067   
113         068   
114         069   
115         069   
116         069   
117         069   
118         070   
119         071   
120         072   
121         073   
122         074   
123         075   
124         076   
125         077   
126         078   
127         079   
128         079   
129         080   
130         081   
131         082   
132         083   
133         083   
134         083   
135         084   
136         085   
137         086   
138         086   
139         087   
140         087   
141         088   
142         089   
143         090   
144         090   
145         091   
146         092   
147         092   
148         092   
149         093   
150         094   
151         094   
152         094   
153         094   
154         094   
155         095   
156         096   
157         097   
158         098   
159         098   
160         099   
161         100   
162         101   
163         102   
164         103   
165         104   
166         105   
167         105   
168         105   
169         106   
170         107   
171         107   
172         107   
173         108   
174         108   
175         109   
176         109   
177         109   
178         110   
179         111   
180         112   
181         113   
182         114   
183         115   
184         116   
185         117   
186         118   
187         118   
188         119   
189         120   
190         121   
191         122   
192         123   
193         124   
194         124   
195         125   
196         125   
197         126   
198         127   
199         128   
200         128   
201         128   
202         128   
203         128   
204         129   
205         129   
206         129   
207         130   
208         131   
209         131

In [58]:
merged[merged.duplicated(subset=["headlineid"], keep=False)]

headlineid  \
3           003   
4           003   
5           003   
6           003   
7           003   
14          010   
15          010   
16          010   
17          010   
18          010   
19          010   
20          010   
21          010   
22          010   
23          011   
24          011   
25          011   
26          011   
34          019   
35          019   
36          019   
37          020   
38          020   
45          027   
46          027   
48          029   
49          029   
51          031   
52          031   
53          032   
54          032   
61          039   
62          039   
63          039   
65          041   
66          041   
69          044   
70          044   
71          044   
72          044   
74          046   
75          046   
77          048   
78          048   
79          048   
86          055   
87          055   
89          057   
90          057   
91          057   
92          057   
93          057   
95          059   
96          059   
98          061   
99          061   
100         061   
102         063   
103         063   
104         063   
105         063   
106         063   
107         064   
108         064   
110         066   
111         066   
114         069   
115         069   
116         069   
117         069   
127         079   
128         079   
132         083   
133         083   
134         083   
137         086   
138         086   
139         087   
140         087   
143         090   
144         090   
146         092   
147         092   
148         092   
150         094   
151         094   
152         094   
153         094   
154         094   
158         098   
159         098   
166         105   
167         105   
168         105   
170         107   
171         107   
172         107   
173         108   
174         108   
175         109   
176         109   
177         109   
186         118   
187         118   
193         124   
194         124   
195         125   
196         125   
199         128   
200         128   
201         128   
202         128   
203         128   
204         129   
205         129   
206         129   
208         131   
209         131   
216         138   
217         138   
219         140   
220         140   
221         140   
222         140   
223         140   
224         140   
225         140   
226         141   
227         141   
228         141   
229         141   
230         142   
231         142   
232         142   
236         146   
237         146   
239         148   
240         148   
241         149   
242         149   
243         149   
244         149   
245         149   
248         152   
249         152   
252         155   
253         155   
254         155   
255         155   
256         155   
260         159   
261         159   
262         159   
263         160   
264         160   
265         160   
266         160   
272         166   
273         166   
275         168   
276         168   
277         168   
278         168   
281         171   
282         171   
283         171   
284         171   
287         174   
288         174   
289         175   
290         175   
291         175   
296         180   
297         180   
298         181   
299         181   
303         185   
304         185   
307         188   
308         188   
309         188   
310         188   
311         188   
312         189   
313         189   
314         189   
315         189   
316         190   
317         190   
320         193   
321         193   
322         193   
323         193   
324         193   
325         193   
326         193   
327         193   
328         193   
329         194   
330         194   
333         197   
334         197   
335         198   
336         198   
337         199   
338         199   
339         200   
340         200   
341         200   
342         201

> 📝 **Note**: We found 176 headlines that were matched with multiple cities the have the same name. For example, There were 3 different **"Rome"** city, one in the Italy and another two in the US. Thus, the headline **"Authorities are Worried about the Spread of Mad Cow Disease in Rome"** got three duplicates.
> What we need to do is to 

### US counties extraction

In [ ]:
df["us_county"] = df.headline.apply(func=lambda x: get_matched_uscounty(x))

print(df.shape)
df.head(10)

### US States extraction

In [ ]:
df["us_state"] = df.headline.apply(func=lambda x: get_matched_usstate(x))
print(df.shape)
df.head(10)


### Countries extraction

In [ ]:
df["country"] = df["headline"].apply(func=lambda x: get_matched_country(x))
print(df.shape)
df.head(10)

In [ ]:
df[df.country.notnull()]

In [51]:
df.merge(
    right=city[["countryname", "cityname", "cityid"]],
    how="left",
    left_on="city",
    right_on="cityname",
    suffixes=("_headline", "_city"),
)

headlineid  \
0           000   
1           001   
2           002   
3           003   
4           003   
5           003   
6           003   
7           003   
8           004   
9           005   
10          006   
11          007   
12          008   
13          009   
14          010   
15          010   
16          010   
17          010   
18          010   
19          010   
20          010   
21          010   
22          010   
23          011   
24          011   
25          011   
26          011   
27          012   
28          013   
29          014   
30          015   
31          016   
32          017   
33          018   
34          019   
35          019   
36          019   
37          020   
38          020   
39          021   
40          022   
41          023   
42          024   
43          025   
44          026   
45          027   
46          027   
47          028   
48          029   
49          029   
50          030   
51          031   
52          031   
53          032   
54          032   
55          033   
56          034   
57          035   
58          036   
59          037   
60          038   
61          039   
62          039   
63          039   
64          040   
65          041   
66          041   
67          042   
68          043   
69          044   
70          044   
71          044   
72          044   
73          045   
74          046   
75          046   
76          047   
77          048   
78          048   
79          048   
80          049   
81          050   
82          051   
83          052   
84          053   
85          054   
86          055   
87          055   
88          056   
89          057   
90          057   
91          057   
92          057   
93          057   
94          058   
95          059   
96          059   
97          060   
98          061   
99          061   
100         061   
101         062   
102         063   
103         063   
104         063   
105         063   
106         063   
107         064   
108         064   
109         065   
110         066   
111         066   
112         067   
113         068   
114         069   
115         069   
116         069   
117         069   
118         070   
119         071   
120         072   
121         073   
122         074   
123         075   
124         076   
125         077   
126         078   
127         079   
128         079   
129         080   
130         081   
131         082   
132         083   
133         083   
134         083   
135         084   
136         085   
137         086   
138         086   
139         087   
140         087   
141         088   
142         089   
143         090   
144         090   
145         091   
146         092   
147         092   
148         092   
149         093   
150         094   
151         094   
152         094   
153         094   
154         094   
155         095   
156         096   
157         097   
158         098   
159         098   
160         099   
161         100   
162         101   
163         102   
164         103   
165         104   
166         105   
167         105   
168         105   
169         106   
170         107   
171         107   
172         107   
173         108   
174         108   
175         109   
176         109   
177         109   
178         110   
179         111   
180         112   
181         113   
182         114   
183         115   
184         116   
185         117   
186         118   
187         118   
188         119   
189         120   
190         121   
191         122   
192         123   
193         124   
194         124   
195         125   
196         125   
197         126   
198         127   
199         128   
200         128   
201         128   
202         128   
203         128   
204         129   
205         129   
206         129   
207         130   
208         131   
209         131

In [ ]:
df

In [ ]:
for ct_idx, ct in enumerate(countries.countryname):
    for st_idx, st in enumerate(df.headline):
        if ct in st:
            print(f"Headline: {st}\t\t\tCountry: {ct}")
            break

In [ ]:
suffix = [
    "County",
    "Municipality",
    "Municipio",
    "Census Area",
    "City and Borough",
    "Borough",
    "Parish",
    "City",
]

for s in suffix:
    print(f"\nSuffix: {s}")
    for headline in df.headline:
        print(f"\t{headline}") if s in headline else None

In [ ]:
city_regexs = []

for city in cities.name:
    #     r1 = '\\b'+city+'\\b'
    #     city_regexs.append(r1)
    r2 = city
    city_regexs.append(r2)
for state in states.name:
    r3 = "\\b" + state + "\\b"
    city_regexs.append(r3)
for county in counties.county:
    r4 = "\\b" + county + "\\b"
    city_regexs.append(r4)

ind = []
city = []
hline = []

for val in [
    "Mala",
    "Bron",
    "Viru",
    "Pati",
    "Rota",
    "Will",
    "Green",
    "\\b" + "Will" + "\\b",
]:
    for regex in city_regexs:
        if regex == val:
            city_regex = city_regexs.remove(val)

for regex in city_regexs:
    compiled_city = re.compile(regex)

    for index, headline in enumerate(df.headline):
        match = compiled_city.search(headline)
        if match is not None:
            start, end = match.start(), match.end()
            matched_string = headline[start:end]
            if len(matched_string) > 3:
                ind.append(index)
                city.append(matched_string)
                hline.append(headline)
#                 print(index,matched_string, '<<<', headline)

# Create dataframe of matched results and sort values by headline_no
matched = {"headline_no": ind, "headline": hline, "city": city}
matched_cities = pd.DataFrame(matched)
matched_cities = matched_cities.sort_values(by="headline_no").reset_index(drop=True)

matched_cities.head(n=15)

In [ ]:
cities[cities.name == "York"]

**Note: We found some interesting patterns of matched results.**

**For example**
* There were **three** matched results on a headline "Could Zika Reach New York City?"
* The first result was "New York City" which was a correct matched. 
* The rest two were "York" which could either be matched but why did we have two "York"?
* Check out on the `cities` dataframe then we found that there are two different locations of "York", one in the US and another one in the GB.

In [ ]:
# How many headline had matched result more than 1?
print("*********" * 10)
print("**** Checked duplicates ****")
print("*********" * 10)
print("\n")
print(
    str(
        len(
            matched_cities.headline.value_counts()[
                matched_cities.headline.value_counts() > 1
            ]
        )
    )
    + "/650 headlines had matched result more than 1."
)

# Drop duplicates of matched city names
matched_cities_uniq = matched_cities.drop_duplicates()

print("\n")
print(
    "Note: There were still number of headlines which had matched result more than 1. However, among those unique matched city names, only the longest string of city is the correct matched"
)
print("\n")

print("*********" * 10)
print("**** Dropped duplicates ****")
print("*********" * 10)
print("\n")
print(
    str(
        len(
            matched_cities_uniq.headline.value_counts()[
                matched_cities_uniq.headline.value_counts() > 1
            ]
        )
    )
    + "/650 headlines had matched result more than 1 after dropped duplicates."
)
print("\n")
print(
    matched_cities_uniq.headline.value_counts()[
        matched_cities_uniq.headline.value_counts() > 1
    ].head(10)
)
print("\n")


# Make a list of headlines that had matched result more than 1.
redun_headlines = matched_cities_uniq.headline.value_counts()[
    matched_cities_uniq.headline.value_counts() > 1
].index.tolist()
# Keep only the longest matched cities
for hl in redun_headlines:
    cities_to_compare = matched_cities_uniq.city[matched_cities_uniq.headline == hl]
    length_str = [
        (len(city), index)
        for index, city in list(zip(cities_to_compare.index, cities_to_compare))
    ]
    for length, index in length_str:
        if (length, index) != max(length_str):
            matched_cities_uniq.drop(index, axis=0, inplace=True)

print("\n")
print("*********" * 10)
print("**** Filtered only the longest matched cities ****")
print("*********" * 10)
print("\n")
print(
    str(
        len(
            matched_cities_uniq.headline.value_counts()[
                matched_cities_uniq.headline.value_counts() > 1
            ]
        )
    )
    + "/650 headlines had matched result more than 1 after filtered the longest."
)
matched_cities_uniq = matched_cities_uniq.reset_index(drop=False).set_index(
    "headline_no", drop=True
)
display(matched_cities_uniq.head())

In [ ]:
#  Add city column to df dataframe
df = df.join(matched_cities_uniq[["city"]])

# Manually fill the nulls in city column.
for i in df[df.city.isnull()].index:
    df.loc[i, "city"] = "Cebu "  # <<<  Zika infects pregnant woman in Cebu
    df.loc[i, "city"] = "Antigua "  # <<<  Spanish Flu Sighted in Antigua
    df.loc[i, "city"] = (
        "Rio De Janeiro"  # <<<  Carnival under threat in Rio De Janeiro due to Zika outbreak
    )
    df.loc[i, "city"] = "Oton"  # <<<  Zika case reported in Oton
    df.loc[i, "city"] = "Maka"  # <<<  Maka City Experiences Influenza Outbreak
    df.loc[i, "city"] = "Mcallen"  # <<<  More Zika patients reported in Mcallen
    df.loc[i, "city"] = (
        "Mclean"  # <<<  More people in Mclean are infected with Hepatitis A every year
    )
    df.loc[i, "city"] = "Sussex"  # <<<  Malaria Exposure in Sussex
    df.loc[i, "city"] = "Greenwich"  # <<<  Greenwich Establishes Zika Task Force
    df.loc[i, "city"] = "Yulee"  # <<<  Yulee takes a hit from Spreading Sickness
    df.loc[i, "city"] = (
        "Boucau"  # <<<  More people in Boucau are infected with HIV every year
    )
    df.loc[i, "city"] = "Manhasset"  # <<<  Bronchitis Outbreak in Manhasset
    df.loc[i, "city"] = "Padre Las Casas"  # <<<  Zika Troubles come to Padre Las Casas
    df.loc[i, "city"] = "Destin"  # <<<  Outbreak of Zika in Destin
    df.loc[i, "city"] = (
        "Gympie"  # <<<  Gympie Patient in Critical Condition after Contracting Chlamydia
    )
    df.loc[i, "city"] = "Druid Hills"  # <<<  Spike of Meningitis Cases in Druid Hills
    df.loc[i, "city"] = (
        "Magnolia"  # <<<  More Patients in Magnolia are Getting Diagnosed with Malaria
    )
    df.loc[i, "city"] = (
        "Penal"  # <<<  Rumors about Syphilis spreading in Penal have been refuted
    )
    df.loc[i, "city"] = "Lisbon"  # <<<  Spanish Flu Outbreak in Lisbon
    df.loc[i, "city"] = "Madrid"  # <<<  Spanish Flu Spreading through Madrid
    df.loc[i, "city"] = "Belvoir"  # <<<  Fort Belvoir tests new cure for Hepatitis C
    df.loc[i, "city"] = (
        "Oak Brook"  # <<<  More people in Oak Brook are infected with Respiratory Syncytial Virus every year
    )
    df.loc[i, "city"] = "Hutchins"  # <<<  Outbreak of Zika in Hutchins
    df.loc[i, "city"] = "Longwood"  # <<<  Longwood volunteers spreading Zika awareness
    df.loc[i, "city"] = "Quixere"  # <<<  Zika symptoms spotted in Quixere
    df.loc[i, "city"] = "Davos"  # <<<  Measles Hits Davos
    df.loc[i, "city"] = (
        "Morehead City"  # <<<  Spike of Hepatitis E Cases in Morehead City
    )
    df.loc[i, "city"] = "Alvorad"  # <<<  Outbreak of Zika in Alvorada
    df.loc[i, "city"] = "Dangriga"  # <<<  Zika arrives in Dangriga
    df.loc[i, "city"] = (
        "Maynard"  # <<<  More Patients in Maynard are Getting Diagnosed with Syphilis
    )
    df.loc[i, "city"] = "Antioquia"  # <<<  Zika case reported in Antioquia
    df.loc[i, "city"] = "Pismo Beach"  # <<<  Chikungunya has not Left Pismo Beach
    df.loc[i, "city"] = "La Joya"  # <<<  Zika spreads to La Joya
#     print("df.loc[i,'city'] = '' #<<< ",df.loc[i,'headline'] ,"\n")
df.info()

In [ ]:
df.head(50)

In [ ]:
# Save to csv file
# df.to_csv("cities_in_headline.csv", index=None)